File: src/gee_sibs_or.c
/**********************************************************************
 * 
 * gee_sibs_or.c
 *
 * Karl W. Broman
 *
 * first written 17 Sept 2003
 * last modified 23 Sept 2003
 *
 * The goal of this is to estimate the log odds ratio between siblings 
 * for some binary response, after controlling for covariates.  We 
 * seek to solve the GEE corresponding to logit{E(y|x)} = x beta
 * and log OR(y1, y2 | x) = gamma.
 *
 * See the following two papers:
 *
 *   KY Liang and TH Beaty (1991) Measuring familial aggregation by
 *   using odds-ratio regression models.  Genet Epidemiol 8:361-370.
 * 
 *   KY Liang, SL Zeger, B Qaqish (1992) Multivariate regression
 *   analyses for categorical data (with discussion). JRSS B
 *   54(1):3-40.
 *
 * The latter paper has all of the details.
 *
 **********************************************************************/

#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <R.h>
#include <Rmath.h>
#include <R_ext/PrtUtil.h>

#include "gee_sibs_or.h"
#include "calc_eta.h"
#include "util.h"

/**********************************************************************
 * 
 * gee_sibs_or_bits
 * 
 * This function is the main function for the essential bits of the 
 * GEE calculation for a single family.  
 *
 * INPUT:
 *     n_sibs = number of siblings in the family
 *     n_covar = number of covariates
 *     n_vecA = n_sibs + choose(n_sibs, 2) = n_sibs + n_sibs*(n_sibs-1)/2
 *     y = binary phenotype (vector of length n_sibs)
 *     x = covariates (matrix of dimension n_covar x n_sibs; note transpose of usual)
 *     beta = current estimates of coefficients (vector of length n_covar)
 *     gamma = current estimate of log OR for siblings
 *
 * OUTPUT:
 *     MatC = d(mu, eta) / dtheta  [a matrix of size (n_covar+1) x n_vecA]
 *     MatBinv = cov(y, w)         [a matrix of size n_vecA x n_vecA] 
 *     vecA = (y-mu, w-eta)        [a vector of length n_vecA]
 *
 *  Note that MatC and MatBinv are indexed as MatC[col][row] col=phe, row=covar
 **********************************************************************/
void gee_sibs_or_bits(int n_sibs, int n_covar, int n_vecA, 
		      int *y, double **X, double *beta, double gamma,
		      double **MatC, double **MatBinv, double *vecA,
		      int maxit, double eta_tol, int trace)
{
  int i, j, k, s, u, v, *w, n_pairs;
  int **pair_index, ***trip_index, ****quad_index;
  double *mu, *eta12=0, *eta123=0, *eta1234=0;

  n_pairs = n_vecA - n_sibs;

  /* Allocate space for the mu's, w's and eta's */
  mu = (double *)R_alloc(n_sibs, sizeof(double));
  w = (int *)R_alloc(n_pairs, sizeof(int)); 
  if(n_sibs > 1)
    eta12 = (double *)R_alloc(n_sibs * (n_sibs - 1) / 2, sizeof(double));
  if(n_sibs > 2)
    eta123 = (double *)R_alloc(n_sibs * (n_sibs - 1) * (n_sibs - 2) / (3*2), 
			       sizeof(double));
  if(n_sibs > 3)
    eta1234 = (double *)R_alloc(n_sibs * (n_sibs - 1) * (n_sibs - 2) * (n_sibs - 3) / 
				(4*3*2), sizeof(double));
  
  /* allocate space for index matrices: simplify indexing eta12, eta123, and eta1234 */
  if(n_sibs > 1)
    allocate_imatrix(&pair_index, n_sibs, n_sibs);
  if(n_sibs > 2)
    allocate_3d_iarray(&trip_index, n_sibs, n_sibs, n_sibs);
  if(n_sibs > 3)
    allocate_4d_iarray(&quad_index, n_sibs, n_sibs, n_sibs, n_sibs);

  /* calculate the fitted values */
  for(i=0; i<n_sibs; i++) {
    mu[i] = 0.0;
    for(j=0; j<n_covar; j++) mu[i] += beta[j] * X[i][j];
    mu[i] = exp(mu[i])/(1.0 + exp(mu[i]));
  }
      
  /* calculate the cross-products, eta12's and pair_index (if n_sibs > 1) */
  for(i=0, u=0; i<n_sibs-1; i++) {
    for(j=i+1; j<n_sibs; j++, u++) {
      w[u] = y[i]*y[j];
      eta12[u] = calc_eta12(mu[i], mu[j], gamma, eta_tol);
      pair_index[i][j] = pair_index[j][i] = u;
    }
  }

  /* calculate the eta123's and trip_index (if n_sibs > 2) */
  for(i=0, u=0; i<n_sibs-2; i++) {
    for(j=i+1; j<n_sibs-1; j++) {
      for(k=j+1; k<n_sibs; k++, u++) {
	eta123[u] = calc_eta123(mu[i], mu[j], mu[k],
				eta12[pair_index[i][j]], 
				eta12[pair_index[i][k]],
				eta12[pair_index[j][k]],
				maxit, eta_tol, trace);
	trip_index[i][j][k] = trip_index[i][k][j] = trip_index[j][i][k] = 
	  trip_index[j][k][i] = trip_index[k][i][j] = trip_index[k][j][i] = u;
      }
    }
  }

  /* calculate the eta1234's and quad_index (if n_sibs > 3) */
  for(i=0, u=0; i<n_sibs-3; i++) {
    for(j=i+1; j<n_sibs-2; j++) {
      for(k=j+1; k<n_sibs-1; k++) {
	for(s=k+1; s<n_sibs; s++, u++) {
	  eta1234[u] = calc_eta1234(mu[i], mu[j], mu[k], mu[s],
				    eta12[pair_index[i][j]], 
				    eta12[pair_index[i][k]],
				    eta12[pair_index[i][s]],
				    eta12[pair_index[j][k]], 
				    eta12[pair_index[j][s]],
				    eta12[pair_index[k][s]],
				    eta123[trip_index[i][j][k]],
				    eta123[trip_index[i][j][s]],
				    eta123[trip_index[i][k][s]],
				    eta123[trip_index[j][k][s]],
				    maxit, eta_tol, trace);
	  quad_index[i][j][k][s] = quad_index[i][j][s][k] = 
	    quad_index[i][k][j][s] = quad_index[i][k][s][j] = 
	    quad_index[i][s][j][k] = quad_index[i][s][k][j] = 
	  quad_index[j][i][k][s] = quad_index[j][i][s][k] = 
	    quad_index[j][k][i][s] = quad_index[j][k][s][i] = 
	    quad_index[j][s][i][k] = quad_index[j][s][k][i] = 
	  quad_index[k][j][i][s] = quad_index[k][j][s][i] = 
	    quad_index[k][i][j][s] = quad_index[k][i][s][j] = 
	    quad_index[k][s][j][i] = quad_index[k][s][i][j] = 
	  quad_index[s][j][k][i] = quad_index[s][j][i][k] = 
	    quad_index[s][k][j][i] = quad_index[s][k][i][j] = 
	    quad_index[s][i][j][k] = quad_index[s][i][k][j] = u;
	}
      }
    }
  }

  /* calculate the vector A */
  for(i=0; i<n_sibs; i++) 
    vecA[i] = (double)y[i] - mu[i];
  for(i=0; i<n_pairs; i++) 
    vecA[n_sibs+i] = (double)w[i] - eta12[i];


  /* Calculate the matrix matBinv = cov(y w) */
  for(i=0; i<n_sibs; i++) { /* first cov(y) */
    MatBinv[i][i] = mu[i]*(1.0-mu[i]);
    for(j=i+1; j<n_sibs; j++) 
      MatBinv[i][j] = MatBinv[j][i] = eta12[pair_index[i][j]] - mu[i]*mu[j];
  }

  for(i=0; i<n_sibs-1; i++) { /* now cov(w) */
    for(j=i+1; j<n_sibs; j++) {
      u = n_sibs + pair_index[i][j]; /* position in MatBinv for first pair */

      for(k=0; k<n_sibs-1; k++) {
	for(s=k+1; s<n_sibs; s++) {

	  v = n_sibs + pair_index[k][s]; /* position in MatBinv for second pair */
    

	  if(i==k && j==s) { /* same pair twice: get variance */
	    MatBinv[u][v] = eta12[pair_index[i][j]]*(1.0-eta12[pair_index[i][j]]);
	  }
	  else if(i==k) { /* pairs (i,j) and (i,s) */
	    MatBinv[u][v] = 
	      eta123[trip_index[i][j][s]] - eta12[pair_index[i][j]] * 
	      eta12[pair_index[i][s]];
	  }
	  else if(i==s) { /* pairs (i,j) and (k,i) */
	    MatBinv[u][v] = 
	      eta123[trip_index[i][j][k]] - eta12[pair_index[i][j]] * 
	      eta12[pair_index[k][i]];
	  }
	  else if(j==k) { /* pairs (i,j) and (j,s) */
	    MatBinv[u][v] = 
	      eta123[trip_index[i][j][s]] - eta12[pair_index[i][j]] *
	      eta12[pair_index[j][s]];
	  }
	  else if(j==s) { /* pairs (i,j) and (k,j) */
	    MatBinv[u][v] = 
	      eta123[trip_index[i][k][j]] - eta12[pair_index[i][j]] *
	      eta12[pair_index[k][j]];
	  }
	  else { /* (i,j) and (k,s) */
	    MatBinv[u][v] = 
	      eta1234[quad_index[i][j][k][s]] - eta12[pair_index[i][j]] *
	      eta12[pair_index[k][s]];
	  }
	}
      }
    }
  }

  for(i=0; i<n_sibs; i++) { /* finally, cov(y, w) */
    for(j=0; j<n_sibs-1; j++) {
      for(k=j+1; k<n_sibs; k++) {
	v = n_sibs + pair_index[j][k];
	
	if(i==j) { /* i and (i,k) */
	  MatBinv[i][v] = MatBinv[v][i] = 
	    eta12[pair_index[i][k]]*(1.0-mu[i]);
	}
	else if(i==k) { /* i and (j,i) */
	  MatBinv[i][v] = MatBinv[v][i] = 
	    eta12[pair_index[j][i]]*(1.0-mu[i]);
	}
	else { /* i and (j,k) */
	  MatBinv[i][v] = MatBinv[v][i] =
	    eta123[trip_index[i][j][k]] - mu[i]*eta12[pair_index[j][k]];
	}
      }
    }
  }


  /* The last bit: MatC = d(mu, eta) / dtheta */
  for(i=0; i<n_sibs; i++) { /* first derivative of fitted values rel to betas */
    for(k=0; k<n_covar; k++) /* note that MatC indexed as MatC[phe][covar] */
      MatC[i][k] = X[i][k]*mu[i]*(1.0-mu[i]);

    MatC[i][n_covar] = 0.0;
  }

  /* now derivative of the eta12's */
  for(i=0; i<n_sibs-1; i++) {
    for(j=i+1; j<n_sibs; j++) {
      u = n_sibs + pair_index[i][j]; 

      /*
      temp1 = 1.0 - (mu[i]+mu[j])*(1.0 - exp(gamma));
      temp2 = sqrt(temp1*temp1 - 4.0*(exp(gamma)-1.0)*exp(gamma)*mu[i]*mu[j]); 
      */

      for(k=0; k<n_covar; k++) {

	/*
	MatC[u][k] = (MatC[i][k] + MatC[j][k])/2.0 - 
	  (2.0*temp1*(exp(gamma)-1)*(MatC[i][k] + MatC[j][k]) - 
	   4.0*(exp(gamma)-1.0)*exp(gamma)*(mu[i]*MatC[j][k] + mu[j]*MatC[i][k])) /
	  (4.0 * (exp(gamma) - 1.0) * temp2); 
	*/

	MatC[u][k] = deta12_dmu(mu[i], mu[j], gamma) * MatC[i][k] +
	  deta12_dmu(mu[j], mu[i], gamma) * MatC[j][k];
      }

      /*
      MatC[u][n_covar] = -exp(gamma)/(2.0*(exp(gamma)-1.0)*(exp(gamma)-1.0)) -
	((2.0*(exp(gamma)-1.0)*(2.0*temp1*(mu[i]+mu[j])*exp(gamma)) -
	  4.0*(2.0*exp(2.0*gamma)-exp(gamma))*mu[i]*mu[j])/(2.0*temp2) - 
	 temp2*2.0*exp(gamma)) / (4.0*(exp(gamma)-1.0)*(exp(gamma)-1.0));
      */

      MatC[u][n_covar] = deta12_dgamma(mu[i], mu[j], gamma);
    }
  }
}


/* wrapper for call from R */
void R_gee_sibs_or_bits(int *n_sibs, int *n_covar, int *n_vecA, 
			int *y, double *x, double *beta, double *gamma,
			double *matC, double *matBinv, double *vecA,
			int *maxit, double *eta_tol, int *trace)
{
  double **X, **MatC, **MatBinv;

  /* reorganize x as a double-indexed matrix, X */
  reorg_dmatrix(x, &X, *n_covar, *n_sibs);

  /* reorganize matC as double-indexed matrix MatC */
  reorg_dmatrix(matC, &MatC, *n_covar+1, *n_vecA);

  /* reorganize matBinv as double-indexed matrix MatBinv */
  reorg_dmatrix(matBinv, &MatBinv, *n_vecA, *n_vecA);

  gee_sibs_or_bits(*n_sibs, *n_covar, *n_vecA, y, X, beta, *gamma,
		   MatC, MatBinv, vecA, *maxit, *eta_tol, *trace);
}


/**********************************************************************
 * deta12_dmu: calculates derivative of eta12 = E(y1*y2) with respect 
 *             to mu1 = E(y1)
 **********************************************************************/
double deta12_dmu(double mu1, double mu2, double gamma)
{
  double exp_gamma, qi_prime, q, ai_prime, a;

  exp_gamma = exp(gamma);

  a = 1.0 + (mu1 + mu2)*(exp_gamma - 1.0);
  q = a*a - 4.0 * exp_gamma * (exp_gamma - 1.0) * mu1 * mu2;

  ai_prime = exp_gamma - 1.0;
  qi_prime = 2.0*a*ai_prime - 4.0 * exp_gamma * (exp_gamma - 1.0) * mu2;

  return( (2.0*ai_prime - qi_prime/sqrt(q))/(4.0*(exp_gamma-1.0)) );
}

/**********************************************************************
 * deta12_dgamma: calculates the derivate of eta12=E(y1*y2) with 
 *                respect to gamma = log OR(y1, y2)
 **********************************************************************/
double deta12_dgamma(double mu1, double mu2, double gamma)
{
  double exp_gamma, qij_prime, q, aij_prime, a;

  exp_gamma = exp(gamma);

  a = 1.0 + (mu1 + mu2)*(exp_gamma - 1.0);
  q = a*a - 4.0 * exp_gamma * (exp_gamma - 1.0) * mu1 * mu2;

  aij_prime = (mu1 + mu2);
  qij_prime = 2.0*a*aij_prime - 4.0 * (2.0*exp_gamma - 1.0) * mu1 * mu2;

  return( exp_gamma * ((aij_prime - 0.5*qij_prime/sqrt(q))*(exp_gamma-1.0) - (a - sqrt(q))) /
	  (2.0*(exp_gamma-1.0)*(exp_gamma-1.0)) );
}


/* end of gee_sibs_or.c */

File: R/gee_sibs_or.R
# gee.sibs.or
#
#' Estimate odds ratio for siblings adjusting for covariates
#'
#' Use GEE to estimate a common sib-sib odds ratio for a binary phenotype,
#' adjusting for covariates with a logistic model.
#'
#' We assume a set of randomly ascertained sibships with measurements on a
#' binary phenotype and a set of covariates.  We use a logistic model to
#' contact the expected phenotype and the covariates and assume a constant log
#' odds ratio for sibs' phenotypes.  Parameters are estimated using generalized
#' estimating equations (GEE).
#'
#' The idea is described in Liange and Beaty (1991).  All of the details appear
#' in Liang et al. (1992) and Qaqish (1990).
#'
#' Key bits are coded in C; but there's still a lot done directly in R.
#' Eventually I'd like to move more of the code to C, for the sake of speed.
#'
#' @param y A vector of binary phenotypes (coded 0/1).
#' @param x A numeric matrix of covariates, including an intercept term.
#' @param id A vector of integers indicating family/group assignment.
#' @param beta Optional starting values for the covariate coefficients; if
#' `length(beta) = ncol(x)+1`, the last value is assumed to be a starting
#' value for `gamma`.
#' @param gamma Optional starting value for the log odds ratio.  If
#' `gamma=0`, it is assumed to be 0 and is not estimated.
#' @param give.se If true, calculate estimated standard errors.
#' @param return.intercept If TRUE, the estimate for the intercept coefficient
#' is included in the output.
#' @param maxit Maximum number of iterations.  If 0, some debugging information
#' is returned.
#' @param tol Tolerance value for determining convergence in Newton's method to
#' solve the GEE.
#' @param eta.tol Tolerance value for determining convergence in Newton's
#' method to calculate the eta parameters.
#' @param trace Indicates whether to display tracing information; large
#' indicators result in more verbose output.
#' @param debug Indicates whether to display special debugging-related
#' information.
#' @param method Indicates whether to use GEE2, GEE1, or the identity matrix as
#' the working covariance matrix.
#'
#' @return If `give.se=TRUE`, the output is a matrix with four columns:
#' the parameter estimates, estimated SEs, \eqn{(est/SE)^2}, and P-values.  If
#' `give.se=FALSE`, a vector with only the parameter estimates is given.
#'
#' The first item is the sib-sib log odds ratio; the rest are the covariates'
#' coefficients; if `return.intercept=FALSE`, the intercept is excluded
#' from the output.
#'
#' @author Karl W Broman, \email{broman@@wisc.edu}
#'
#' @seealso [fake.data()]
#'
#' @references Liang, K.-Y. and Beaty, T. H. (1991) Measuring familial
#' aggregation by using odds-ratio regression models.  *Genetic
#' Epidemiology*, **8**, 361--370.
#'
#' Liang, K.-Y., Zeger, S. L. and Qaqish, B. (1992) Multivariate regression
#' analyses for categorical data (with discussion).  *J. Roy. Statist.
#' Soc.* B, **1**, 3--40.
#'
#' Qaqish, B. F. (1990) Multivariate regression models using generalized
#' estimating equations.  Ph. D. Thesis, Department of Biostatistics, Johns
#' Hopkins University, Baltimore, Maryland.
#'
#' @keywords models
#'
#' @examples
#' data(fake.data)
#' y <- fake.data[,1]
#' id <- fake.data[,2]
#' x <- cbind(1, fake.data[,-(1:2)])
#' gee.output <- gee.sibs.or(y, x, id, trace=TRUE)
#'
#' @importFrom stats glm binomial pchisq
#' @export gee.sibs.or
gee.sibs.or <-
function(y, x, id, beta=NULL, gamma=0.5, give.se=TRUE, return.intercept=FALSE,
         maxit=1000, tol=1e-5, eta.tol=1e-9, trace=FALSE, debug=FALSE,
         method=c("gee2","gee1","identity"))
{
  # check arguments
  if(!is.null(beta) && length(beta) != ncol(x)) {
    if(length(beta) == ncol(x)+1) {
      gamma <- beta[length(beta)]
      beta <- beta[-length(beta)]
    }
    else stop("length(beta) != ncol(x)")
  }
  if(length(y) != nrow(x)) stop("length(y) != nrow(x)")
  if(length(y) != length(id)) stop("length(y) != length(id)")
  if(any(y!=0 & y !=1)) stop("y should have values 0 and 1")
  method <- match.arg(method)

  if(!maxit) eta.maxit <- 1000
  else eta.maxit <- maxit

  # split id into family categories
  wh <- split(1:length(id), id)
  n.fam <- length(wh)
  n.covar <- ncol(x)

  # starting values
  if(is.null(beta))
    beta <- glm(y ~ -1 + x, family=binomial(link=logit))$coef

  names(beta) <- names(gamma) <- NULL
  if(trace) cat(0,beta,gamma,"\n")

  if(!gamma) est.gamma <- FALSE
  else est.gamma <- TRUE

  flag <- 0
  prev.val <- Inf

  result <- vector("list",n.fam)
  for(i in 1:maxit) {
    for(j in 1:n.fam) {
      result[[j]] <- gee.sibs.or.bits(y[wh[[j]]], x[wh[[j]],,drop=FALSE],
                                      beta, gamma, eta.maxit, eta.tol, trace)

      if(any(is.na(unlist(result[[j]]))))
        warning("gee.sibs.or.bits returned NAs", j, "\n")

      len <- length(wh[[j]])
      if(!est.gamma) { # assume gamma=0 (don't estimate it)
        result[[j]]$matC <- result[[j]]$matC[1:n.covar,1:len,drop=FALSE]
        result[[j]]$matBinv <- result[[j]]$matBinv[1:len,1:len,drop=FALSE]
        result[[j]]$vecA <- result[[j]]$vecA[1:len]
      }

      if(method=="gee2") temp <- result[[j]]$matC %*% solve(result[[j]]$matBinv)
      else if(method=="gee1") {
        result[[j]]$matC[1:n.covar,-(1:len)] <-
          result[[j]]$matC[-(1:n.covar),1:len] <- 0
        result[[j]]$matBinv[1:len,-(1:len)] <-
          result[[j]]$matBinv[-(1:len),1:len] <- 0
        temp <- result[[j]]$matC %*% solve(result[[j]]$matBinv)
      }
      else temp <- result[[j]]$matC # <--- using identity matrix as working covar mat

      if(j==1) {
        first.mat <- temp %*% t(result[[j]]$matC)
        temp2 <- temp %*% result[[j]]$vecA
        second.mat <- temp2
        third.mat <- temp2 %*% t(temp2)
      }
      else {
        first.mat <- first.mat + temp %*% t(result[[j]]$matC)
        temp2 <- temp %*% result[[j]]$vecA
        second.mat <- second.mat + temp2
        third.mat <- third.mat + temp2 %*% t(temp2)
      }
    }
    if(!maxit) return(first.mat,second.mat)

    step <- solve(first.mat, second.mat)

    cur.val <- max(abs(second.mat))
#    if(cur.val > prev.val) {
#      step <- step/32
#      cat("  ** step/32 **\n")
#    }

    if(est.gamma) {
      newbeta <- beta + step[-length(step)]
#      if(cur.val <= prev.val)
        newgamma <- gamma + step[length(step)]
#      else
#        newgamma <- gamma + step[length(step)]
    }
    else {
      newbeta <- beta + step
      newgamma <- gamma
    }

    prev.val <- cur.val

    if(all(abs(beta-newbeta)<tol) & abs(gamma-newgamma)<tol) {
      flag <- 1
      break
    }
    beta <- newbeta; gamma <- newgamma

    if(debug) {
      z <- as.numeric(second.mat)
      u <- rev(order(abs(z)))
      z <- z[u][1:min(c(length(z),3))]
      cat(i,z,gamma,"\n")
    }
    if(trace) cat(i,beta,gamma,"\n")
  }
  if(!flag) warning(" -Didn't converge.\n")

  if(trace>1) {
    cat("\n")
    cat(as.numeric(second.mat),"\n")
    cat("\n")
  }

  output <- c(gamma,beta)

  if(give.se) {
    first.mat <- solve(first.mat)
    var.mat <- first.mat %*% third.mat %*% t(first.mat)
    se <- sqrt(diag(var.mat))
    se <- c(se[length(se)],se[-length(se)]) # reorder so ln(OR) is first

    output <- cbind(output, se, (output/se)^2)
    output <- cbind(output, 1-pchisq(output[,3],1))

    colnames(output) <- c("est","se","Wald","P")
    rownames(output) <- c("lnOR",colnames(x))
    if(!return.intercept) output <- output[-2,,drop=FALSE]
  }
  else {
    names(output) <- c("lnOR",colnames(x))
    if(!return.intercept) output <- output[-2]
  }

  output
}


#' @useDynLib geesibsor, .registration=TRUE

gee.sibs.or.bits <-
function(y, x, beta, gamma, maxit=1000,
         eta.tol=1e-9, trace=FALSE)
{
  n.sibs <- length(y)
  n.covar <- ncol(x)
  if(nrow(x) != n.sibs) stop("nrow(x) != length(y)")
  if(ncol(x) != length(beta)) stop("ncol(x) != length(beta)")

  if(n.sibs > 1) n.vecA <- n.sibs + choose(n.sibs,2)
  else n.vecA <- n.sibs

  result <- .C("R_gee_sibs_or_bits",
               as.integer(n.sibs),
               as.integer(n.covar),
               as.integer(n.vecA),
               as.integer(y),
               as.double(t(x)), # input transpose of covariate matrix
               as.double(beta),
               as.double(gamma),
               matC = as.double(rep(0,(n.covar+1)*n.vecA)),
               matBinv = as.double(rep(0,n.vecA*n.vecA)),
               vecA = as.double(rep(0,n.vecA)),
               as.integer(maxit),
               as.double(eta.tol),
               as.integer(trace),
               PACKAGE="geesibsor")

  return(list(matC=matrix(result$matC, nrow=n.covar+1),
              matBinv=matrix(result$matBinv, nrow=n.vecA),
              vecA=result$vecA))
}
